core/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java

```diff
@@ -84,6 +85,7 @@ public GatewayMetaState(Settings settings, NodeEnvironment nodeEnv, MetaStateSer
         if (DiscoveryNode.masterNode(settings) || DiscoveryNode.dataNode(settings)) {
             try {
                 ensureNoPre019State();
+                IndexFolderUpgrader.upgradeIndicesIfNeeded(settings, nodeEnv);
                 pre20Upgrade();
```

Comment:
> I think we should check for too old versions before starting to mess up with folders.

Agreed, now we upgrade index folders after we ensure version of indices are supported